db/range_del_aggregator.cc

```diff
@@ -120,17 +143,24 @@ class UncollapsedRangeDelMap : public RangeDelMap {
 // is installed to indicate that no tombstone exists. This occurs at keys n and
 // t in the example above.
 //
+// To provide end-key-inclusive semantics for truncated range tombstones, the
+// map actually stores two sequence numbers: one for the "point" seqnum that applies
+// to the key of the entry, and one for the "range" seqnum that applies from the
+// current key until the next key, exclusive on both ends. For untruncated range
+// tombstones, the point and range seqnums are equal, so the semantics match
+// to the description above.
```

Comment:
This comment is helpful, though I still haven't quite internalized this representation. On my commute home I was wondering about another representation: represent the start and end key for tombstones with `InternalKeys` rather than with user-keys. For an untruncated tombstone, the start key would be `<start-user-key>@kMaxSequenceNumber` and the end key be `<end-user-key>@kMaxSequenceNumber`. Both the start key and end key would be inclusive. For truncated tombstones, the sequence number would allow allow truncation of the tombstone in the middle of a user-key which would allow us to always truncate to exactly the sstable boundaries (there might be some off-by-1 adjustment on the upper boundary). I think this representation would allow something very similar to the existing `CollapsedRangeDelMap` structure: a `multiset<InternalRangeTombstone>` where the tombstones span internal-keys. The `ShouldDelete` check would require finding the tombstone the key is contained within (pretty much exactly like the existing code, except `parsed.sequence` is also considered), and then verify that the tombstone sequence number is newer than the target sequence number (exactly like the existing code).

Hopefully I'm not derailing the work here too much. I need to think harder about the representation being proposed in this PR.

Reply:
I think this sounds like a promising idea. One possible subtlety with this approach is that we need to take care to avoid gaps between truncated tombstones. Suppose we had SST1 with smallest = a@15 and largest = c@30, and SST2 with smallest = c@20 and largest = e@10. If we had a range tombstone [a, d)@9, we would have to be careful with how we truncate it. If we truncate it to [a@15, c@30] in SST1 and [c@20, e@10] in SST2 (ignoring the possible off-by-1 adjustment you mentioned), then if we had a Get on snapshot 10 that was calling ShouldDelete on a@5, we'd erroneously declare that as existing when it should be deleted.

I think that scenario above can be fixed fairly easily by not truncating the start key for the first SST in a level (i.e., pass nullptr to `smallest`), not truncating the end key for the last SST in a level (i.e., pass nullptr to `largest`), and truncating the end key at the smallest key in the next SST as we've been doing already. However, there could be worse cases than this, so I'd want to think a bit more about your proposed approach before making a switch.

Reply:
> Suppose we had SST1 with smallest = a@15 and largest = c@30, and SST2 with smallest = c@20 and largest = e@10. If we had a range tombstone [a, d)@9, we would have to be careful with how we truncate it. If we truncate it to [a@15, c@30] in SST1 and [c@20, e@10] in SST2 (ignoring the possible off-by-1 adjustment you mentioned), then if we had a Get on snapshot 10 that was calling ShouldDelete on a@5, we'd erroneously declare that as existing when it should be deleted.

I'm not seeing the problem here. Can you elaborate? If you truncate `[a,d)@9` to SST1 you'd get `[a@15,c@30]@9`. A snapshot at 10 would definitely see this tombstone and it covers the key `a@5` (`a@15` <= `a@5` <= `c@30`). 

I do see another possible problem with this approach. With your example, what happens to `c@15` if there were a tombstone `[a,d)@16`? That key is covered by the tombstone `[a,d)@16`, but it lies between SST1 and SST2 and thus the truncated tombstones wouldn't cover it. I think this is ok because it should be impossible for `c@15` to lie at a lower level in the LSM. Subtle .

Reply:
My bad about that example; you're correct that it wouldn't be a problem.

However, I still feel somewhat disinclined to apply this approach to this PR, because I think it's mostly functional in its current form (I still need to write a few more detailed tests with multiple `AddTombstones` calls but I don't think there are any insurmountable issues remaining). Internal-key-based truncation is definitely an interesting idea, and I wonder if this approach can be used in the v2 range tombstone design in some way if it doesn't end up making it here.

Reply:
Understood that the approach in this PR is almost working, though that isn't a strong reason to stick with it if a better approach presents itself. That isn't to say my proposal is better, just that we should be evaluating the proposals fairly, not based on which we got started working on first.

My morning thought about using internal-keys for the tombstone endpoints: we can leave the upper-bound of the tombstone exclusive. Consider the example above with SST1 having bounds `[a@15,c@30]` and SST2 having bounds `[c@0,e@10]`. The tombstone `[a,d)@9` could be truncated by SST1 to `[a@15,c@29)@9`. That is, we use the largest key to truncate the upper bound of the tombstone, but decrement the sequence number by 1. I believe this is safe because we'll never have a tombstone which straddles to sstables where the largest key has a sequence number of 0. If the largest key had a sequence number of 0, the next sstable cannot possibly contain the same user-key. And when the tombstone upper bound is the largest key in an sstable we place the tombstone sentinel key as the sstable largest key.

Now its time for me to grok the approach used in this PR better.

Reply:
> ...we should be evaluating the proposals fairly, not based on which we got started working on first.

That's fair; sunk cost fallacy on my part, which of course isn't really a good reason.

I've finished writing tests so I'm going to give your proposal a try and see if that simplifies things.